File: DFS/dfs_.cpp
#include <iostream>
#include <map>
#include <vector>
using namespace std;

class Graph {
public:
    int V;
    map<int, vector<int>> adj_list;
    bool visited[10000];

    Graph(int n) {
        V = n;
        for(int i=0; i<n; i++) {
            visited[i] = false;
        };
    }

    void addEdge(int u, int v) {
        adj_list[u].push_back(v);
    }

    void DFS_(int at) {
        if(visited[at-1] == true) return;
        visited[at-1] = true;
        cout << at << " ";

        vector<int> neighbors;
        neighbors = adj_list[at];
        for (auto next: neighbors) {
            DFS_(next);
        }
    }

    void displayAdjList() {
        for(auto it: adj_list) {
            cout << it.first << " -> ";
            for(auto x : it.second) {
                cout << x << " ";
            }
            cout << endl;
        }
    }

};

int main() {
    int n, e, u, v;
    cin >> n >> e;

    Graph g(n);

    for (int i=0; i<e; i++) {
        cin >> u >> v;
        g.addEdge(u, v);
    }
    g.displayAdjList();
    g.DFS_(1);

    return 0;
}

File: DFS/dfs_.py
from collections import defaultdict

class Graph:
    def __init__(self, n):
        self.nodes = n
        self.adj_list = defaultdict(list)
        self.visited = n * [False]

    def addEdge(self, u, v):
        self.adj_list[u].append(v)

    def DFS_(self, at):
        if self.visited[at-1]: return
        self.visited[at-1] = True
        print(at, end=' ')

        neighbors = self.adj_list[at]
        for next in neighbors:
            self.DFS_(next)


def main():
    g = Graph(6)
    g.addEdge(1, 2)
    g.addEdge(1, 3)
    g.addEdge(2, 4)
    g.addEdge(2, 5)
    g.addEdge(3, 5)
    g.addEdge(3, 1)
    g.addEdge(4, 5)
    g.addEdge(4, 6)
    g.addEdge(5, 6)
    g.addEdge(5, 3)
    g.addEdge(6, 4)
    g.addEdge(6, 5)
    # print(g.adj_list)

    g.DFS_(1)

if __name__ == '__main__':
    main()
